pkg/batch/executor.go

```diff
@@ -0,0 +1,87 @@
+package batch
+
+import (
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+type BatchFn func() (interface{}, error)
+
+type Batcher interface {
+	BatchFor(key string, dur time.Duration, fn BatchFn) (interface{}, error)
+}
+
+type response struct {
+	v   interface{}
+	err error
+}
+
+type request struct {
+	key              string
+	dur              time.Duration
+	fn               BatchFn
+	responseCallback chan *response
+}
+
+type Executor struct {
+	requests chan *request
+	execs    chan string
+	keys     map[string][]*request
+	logger   logging.Logger
+}
+
+func NewExecutor(logger logging.Logger) *Executor {
+	e := &Executor{
+		requests: make(chan *request),
+		execs:    make(chan string),
+		keys:     make(map[string][]*request),
+		logger:   logger,
+	}
+	go e.Run() // TODO(ozkatz): should probably be managed by the user (also, allow stopping it)
```

Comment:
You can move this one into Run() with defer close, so we will not require to add Stop() and/or Close() methods to handle this resource.